File: leetcode_by_cpp/383.赎金信.cpp
/*
 * @lc app=leetcode.cn id=383 lang=cpp
 *
 * [383] 赎金信
 *
 * https://leetcode-cn.com/problems/ransom-note/description/
 *
 * algorithms
 * Easy (54.29%)
 * Likes:    110
 * Dislikes: 0
 * Total Accepted:    27.5K
 * Total Submissions: 50.5K
 * Testcase Example:  '"a"\n"b"'
 *
 * 给定一个赎金信 (ransom) 字符串和一个杂志(magazine)字符串，判断第一个字符串 ransom 能不能由第二个字符串 magazines
 * 里面的字符构成。如果可以构成，返回 true ；否则返回 false。
 * 
 * (题目说明：为了不暴露赎金信字迹，要从杂志上搜索各个需要的字母，组成单词来表达意思。杂志字符串中的每个字符只能在赎金信字符串中使用一次。)
 * 
 * 
 * 
 * 注意：
 * 
 * 你可以假设两个字符串均只含有小写字母。
 * 
 * canConstruct("a", "b") -> false
 * canConstruct("aa", "ab") -> false
 * canConstruct("aa", "aab") -> true
 * 
 * 
 */

// @lc code=start
#include <unordered_map> 
using namespace std;
class Solution {
private:
    unordered_map<char,int> mapMagazine;
public:
    bool canConstruct(string ransomNote, string magazine) {
        for(auto c : magazine)
            mapMagazine[c]++;
        for(auto c : ransomNote)
        {
            mapMagazine[c]--;
            if(mapMagazine[c] < 0)
                return false;
        }
        return true;
    }
};
// @lc code=end

File: leetcode_by_rust/src/bin/383.赎金信.rs
/*
 * @lc app=leetcode.cn id=383 lang=rust
 *
 * [383] 赎金信
 *
 * https://leetcode.cn/problems/ransom-note/description/
 *
 * algorithms
 * Easy (62.14%)
 * Likes:    855
 * Dislikes: 0
 * Total Accepted:    425.1K
 * Total Submissions: 677.5K
 * Testcase Example:  '"a"\n"b"'
 *
 * 给你两个字符串：ransomNote 和 magazine ，判断 ransomNote 能不能由 magazine 里面的字符构成。
 * 
 * 如果可以，返回 true ；否则返回 false 。
 * 
 * magazine 中的每个字符只能在 ransomNote 中使用一次。
 * 
 * 
 * 
 * 示例 1：
 * 
 * 
 * 输入：ransomNote = "a", magazine = "b"
 * 输出：false
 * 
 * 
 * 示例 2：
 * 
 * 
 * 输入：ransomNote = "aa", magazine = "ab"
 * 输出：false
 * 
 * 
 * 示例 3：
 * 
 * 
 * 输入：ransomNote = "aa", magazine = "aab"
 * 输出：true
 * 
 * 
 * 
 * 
 * 提示：
 * 
 * 
 * 1 <= ransomNote.length, magazine.length <= 10^5
 * ransomNote 和 magazine 由小写英文字母组成
 * 
 * 
 */

// @lc code=start
impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        // let mut ransom_note = ransom_note;
        // let mut magazine = magazine;
        // for c in ransom_note.chars() {
        //     if let Some(pos) = magazine.find(c) {
        //         magazine.remove(pos);
        //     } else {
        //         return false;
        //     }
        // }
        // true
        let mut dict = std::collections::HashMap::new();
        
        for c in magazine.chars() {
            dict.entry(c).and_modify(|count| *count += 1).or_insert(1);
        }
        
        for c in ransom_note.chars() {
            match dict.get_mut(&c) {
                Some(n) if *n > 0 => { *n -= 1; }
                _ => { return false; }
            }
        }
        
        true
    }
}
// @lc code=end

